core/circuitbreaker/rule.go

```diff
@@ -35,307 +33,93 @@ func (s Strategy) String() string {
 	}
 }
 
-// Rule is the base interface of the circuit breaker rule.
-type Rule interface {
-	base.SentinelRule
-	// BreakerStrategy returns the circuit breaker strategy.
-	BreakerStrategy() Strategy
-	// IsApplicable checks whether the rule is valid and could be converted to a corresponding circuit breaker.
-	IsApplicable() error
-	// BreakerStatIntervalMs returns the statistic interval of circuit breaker (in milliseconds).
-	BreakerStatIntervalMs() uint32
-	// IsEqualsTo checks whether current rule is equal to the given rule.
-	IsEqualsTo(r Rule) bool
-	// IsStatReusable checks whether current rule is "statistically" equal to the given rule.
-	IsStatReusable(r Rule) bool
-}
-
-// RuleBase encompasses the common fields of circuit breaking rule.
-type RuleBase struct {
+// Rule encompasses the fields of circuit breaking rule.
+type Rule struct {
 	// unique id
-	Id string
+	Id string `json:"id,omitempty"`
 	// resource name
-	Resource string
-	Strategy Strategy
+	Resource string   `json:"resource"`
+	Strategy Strategy `json:"strategy"`
 	// RetryTimeoutMs represents recovery timeout (in milliseconds) before the circuit breaker opens.
 	// During the open period, no requests are permitted until the timeout has elapsed.
 	// After that, the circuit breaker will transform to half-open state for trying a few "trial" requests.
-	RetryTimeoutMs uint32
+	RetryTimeoutMs uint32 `json:"retryTimeoutMs"`
 	// MinRequestAmount represents the minimum number of requests (in an active statistic time span)
 	// that can trigger circuit breaking.
-	MinRequestAmount uint64
+	MinRequestAmount uint64 `json:"minRequestAmount"`
 	// StatIntervalMs represents statistic time interval of the internal circuit breaker (in ms).
-	StatIntervalMs uint32
+	StatIntervalMs uint32 `json:"statIntervalMs"`
+	// MaxAllowedRtMs indicates that any invocation whose response time exceeds this value (in ms)
+	// will be recorded as a slow request.
+	// MaxAllowedRtMs only takes effect for SlowRequestRatio strategy
+	MaxAllowedRtMs uint64 `json:"maxAllowedRtMs"`
+	// Threshold represents the threshold of circuit breaker.
+	// for SlowRequestRatio, it represents the max slow request ratio
+	// for ErrorRatio, it represents the max error request ratio
+	// for ErrorCount, it represents the max error request count
+	Threshold float64 `json:"threshold"`
 }
 
-func (b *RuleBase) BreakerStatIntervalMs() uint32 {
-	return b.StatIntervalMs
+func (r *Rule) String() string {
+	// fallback string
+	return fmt.Sprintf("{id=%s,resource=%s, strategy=%s, RetryTimeoutMs=%d, MinRequestAmount=%d, StatIntervalMs=%d, MaxAllowedRtMs=%d, Threshold=%f}",
+		r.Id, r.Resource, r.Strategy, r.RetryTimeoutMs, r.MinRequestAmount, r.StatIntervalMs, r.MaxAllowedRtMs, r.Threshold)
 }
 
-func (b *RuleBase) IsApplicable() error {
-	if len(b.Resource) == 0 {
+func (r *Rule) isApplicable() error {
```

Comment:
Export this function is user-friendly?
How do you think of it?

Reply:
Since the rule itself is not an interface anymore, maybe we could make it an exported util checking function (like `IsValidFlowRule` in `flow/rule_manager.go`) instead of a member function?

Reply:
make sense